types/wlr_data_device.c

```diff
@@ -866,10 +866,10 @@ static void data_source_destroy(struct wl_client *client,
 
 static void data_source_set_actions(struct wl_client *client,
 		struct wl_resource *resource, uint32_t dnd_actions) {
-	struct wlr_data_source *source =
+	struct client_data_source *source =
 		wl_resource_get_user_data(resource);
 
-	if (source->actions_set) {
+	if (source->source.actions >= 0) {
```

Comment:
Can this miss the error if actions are "set" to zero?

Reply:
No. The default value is -1 (it's set in `wlr_data_source_init`).